### PR TITLE
Resolve #1022, Resolve #1095 -- The beginning of the AI rewrite

### LINKED ISSUE #1022: Minions continue to follow you.
Minions will following you all through jungle and through to other lanes.
if you go in a bush it will still following you, Essentially you have to kill the minions.

### LINKED ISSUE #1095: Minions Attack Nexus Before Turrets Are Destroyed
Minions seem to ignore the protection system and attack even when they shouldn't.

### PR DESCRIPTION
**Pull request by Killfrra**
The script has been rewritten, implementing the most key points from the article ["Minion AI documentation"](https://proworthy.gg/blog/minion-ai-documentaiton/).
Behavior should now actually be checked every 250ms.
Minions are now able to hold a target outside of their attack radius, but within detection radius. Especially noticeable with melee minions.

### CODE REVIEW

**Review by lzardy — CHANGES_REQUESTED** [review]
Needs a cleanup.
Any un-implemented or unused comments should be removed. Any TODOs should be implemented, then removed as well.
EX:
```
// TODO: Fix the code stupidly copied from the previous version of the script so that it actually does what is written in the comments.
```

File: GameServerCore/Domain/GameObjects/IObjAiBase.cs [hunk 1]
```diff
@@ -70,6 +70,29 @@ public interface IObjAiBase : IAttackableUnit
         /// Unit this AI will auto attack when it is in auto attack range.
         /// </summary>
         IAttackableUnit TargetUnit { get; set; }
+        /// <summary>
+        /// A unit's acquisition range is the distance at which they will automatically acquire attack targets, referred to as 'auto-attacking'.
+        /// </summary>
+        int AcquisitionRange { get; set; }
```
Comment (lzardy) [inline GameServerCore/Domain/GameObjects/IObjAiBase.cs]: Move this to AIScriptMetadata.

File: GameServerCore/Domain/GameObjects/IObjAiBase.cs [hunk 2]
```diff
@@ -70,6 +70,29 @@ public interface IObjAiBase : IAttackableUnit
...
+        /// <summary>
+        /// AIScript should set this to true if it can handle incoming calls for help.
+        /// To have callers populate a list of units attacking allies.
+        /// </summary>
+        bool HandlesCallsForHelp { get; set; }
```
Comment (lzardy) [inline GameServerCore/Domain/GameObjects/IObjAiBase.cs]: Move this to AIScriptMetadata.

File: GameServerLib/GameObjects/AttackableUnits/AI/ObjAIBase.cs [hunk 3]
```diff
@@ -967,6 +966,47 @@ public override void Update(float diff)
             }
         }
 
+        public override void TakeDamage(IAttackableUnit attacker, float damage, DamageType type, DamageSource source, DamageResultType damageText)
+        {
+            base.TakeDamage(attacker, damage, type, source, damageText);
+            OnTakeDamage(attacker);
+        }
+        void OnTakeDamage(IAttackableUnit attacker)
```
Comment (lzardy) [inline GameServerLib/GameObjects/AttackableUnits/AI/ObjAIBase.cs:976]: Remove the unnecessary extra TakeDamage function and place the body inside the TakeDamage above.

File: GameServerCore/Domain/GameObjects/IObjAiBase.cs [hunk 4]
```diff
@@ -70,6 +70,29 @@ public interface IObjAiBase : IAttackableUnit
...
+        /// <summary>
+        /// List of attacking allied units that are asking for help and their priorities.
+        /// </summary>
+        Dictionary<IAttackableUnit, int> unitsAttackingAllies { get; }
```
Comment (lzardy) [inline GameServerCore/Domain/GameObjects/IObjAiBase.cs]: Move this to AIScript (not metadata).

File: GameServerCore/Domain/GameObjects/IObjAiBase.cs [hunk 5]
```diff
@@ -70,6 +70,29 @@ public interface IObjAiBase : IAttackableUnit
...
+        /// <summary>
+        /// Function that populates a list of units attacking allies.
+        /// </summary>
```
Comment (lzardy) [inline GameServerCore/Domain/GameObjects/IObjAiBase.cs]:
```suggestion
        /// </summary>
        /// TODO: Move this to a dedicated AIScriptHandler
```

File: GameServerCore/Domain/GameObjects/IObjAiBase.cs [hunk 6]
```diff
@@ -70,6 +70,29 @@ public interface IObjAiBase : IAttackableUnit
...
+        void CallForHelp(IAttackableUnit attacker, IAttackableUnit victium = null);
+        /// <summary>
+        /// Function that clears a list of units attacking allies.
+        /// It should be called after the list has been processed and is no longer needed to avoid its uncontrolled growth.
+        /// </summary>
```
Comment (lzardy) [inline GameServerCore/Domain/GameObjects/IObjAiBase.cs]:
```suggestion
        /// </summary>
        /// TODO: Move this to a dedicated AIScriptHandler
```

File: GameServerLib/GameObjects/AttackableUnits/AI/ObjAIBase.cs [hunk 7]
```diff
@@ -967,6 +966,47 @@ public override void Update(float diff)
...
+                    && Vector2.DistanceSquared(u.Position, Position) <= (AcquisitionRange * AcquisitionRange)
+                    && Vector2.DistanceSquared(u.Position, attacker.Position) <= (AcquisitionRange * AcquisitionRange)
+                )
+                {
+                    u.CallForHelp(attacker, this);
+                }
+            }
+        }
+
+        public bool HandlesCallsForHelp { get; set; } = false;
```
Comment (lzardy) [inline GameServerLib/GameObjects/AttackableUnits/AI/ObjAIBase.cs]: General style rule is that variables are declared at the top of the class.

**Review by lzardy — CHANGES_REQUESTED** [review]
Changes needed.

File: GameServerCore/Domain/GameObjects/IObjAiBase.cs (hunk 1 above)
Comment (Killfrra) [reply GameServerCore/Domain/GameObjects/IObjAiBase.cs]: Moved to `Stats`, `StatsModifier`, and `CharData`

File: GameServerLib/GameObjects/AttackableUnits/AI/ObjAIBase.cs (hunk 3 above)
Comment (Killfrra) [reply GameServerLib/GameObjects/AttackableUnits/AI/ObjAIBase.cs:976]: Not yet possible, because a second `TakeDamage` function has appeared

**Comment by Killfrra** [comment]
Should solve https://github.com/LeagueSandbox/GameServer/issues/1022 because the chase timeout and calls for help were implemented. Also, the minion loses the target if it ceases to be visible to his team, for example, goes into the bushes. Perhaps later I will need to implement a feature so that it goes to the point where it saw the target for the last time.
Should solve https://github.com/LeagueSandbox/GameServer/issues/1095 because the reading and correct use of the AcquisitionRange stat was implemented, as a result of which the acquisition radius of ranged and melee minions increased (by 47% 😏). In addition, due to calls for help, they can no longer ignore the tower attacking allied unit.

**Review by lzardy — APPROVED** [review]
Good to go. Not entirely finished, but certainly an improvement over the previous implementation.

**Pull request merged**